src/api/web3.ts

```diff
@@ -0,0 +1,210 @@
+        ethereum: provider & {
+            enable: () => Promise<string[]>
+            on?: () => EventListener
+        }
+        web3: Web3
+    }
+}
+
+export interface Web3FrontendAPI {
+    web3: Web3
+    getCurrentAccount: () => Promise<string>
+    getAccounts: () => Promise<string[]>
+    getBalance: (account?: string) => Promise<string>
+    getBlock: (blockNumberOrString: string | number) => Promise<Partial<Block>>
+    getNetwork: () => Promise<string | number>
+    getNetworkName: (id?: number | string) => Promise<string>
+    getNetworkProviderURL: (id: number | string) => string
+    getTransaction: (txHash: string) => Promise<Transaction>
+    getTransactionReceipt: (txHash: string) => Promise<TransactionReceipt>
+    // Utils
+    isAddress: (address: string) => boolean
+    isReadOnly: () => boolean
+    // Provider Specific
+    resetProvider(): void
+    setProvider(provider: string): void
+}
+
+export interface Web3SetupOptions {
+    customProvider?: provider | string
+    reloadOnAccountsChange?: boolean
+}
+
+let web3API: Web3FrontendAPI
+
+let readOnly: boolean
+
+export const getWeb3API = async (options: Web3SetupOptions = {}, force?: boolean): Promise<Web3FrontendAPI> => {
+    if (!force && web3API) return web3API
+
+    return setupWeb3(options)
+}
+
+const getProvider = async ({ customProvider, reloadOnAccountsChange }: Web3SetupOptions): Promise<Web3> => {
+    let provider: Web3
+
+    // wait window load then check window globals...
+    await windowLoaded
+
+    // Use Custom Provider if passed in
+    if (customProvider) provider = new Web3(customProvider)
+    // No Custom Provider? Check window.ethereum
+    else if (window && window.ethereum) {
+        provider = new Web3(window.ethereum)
+
+        // enable Provider, if possible
+        await window.ethereum.enable().catch((error: Error) => {
+            throw new Error(error.message)
+        })
+
+        if (window.ethereum.on && reloadOnAccountsChange) {
+            window.ethereum.on('accountsChanged', function() {
+                window.location.reload()
+            })
+        }
+    }
+    // Check for window.web3's current provider (if it exists)
+    else if (window.web3 && window.web3.currentProvider) provider = new Web3(window.web3.currentProvider)
+    // Else setup localhost version
+    else {
+        try {
+            const url = 'http://localhost:8545'
+            await fetch(url)
+            console.debug('API/WEB3.ts => local node active')
+            provider = new Web3(url)
+        } catch (error) {
+            if (error.readyState === 4 && (error.status === 400 || error.status === 200)) {
+                // the endpoint is active
+                console.debug('API/WEB3.ts catch block => Success')
+            } else {
+                console.debug('API/WEB3.ts =>  No web3 instance injected. Falling back to null provider.')
+                readOnly = true
+                provider = new Web3(null)
```

Comment:
What is a null provider?
Also, what is "readOnly"? If there's no connection with the node you can't read, right?

Reply:
App loads, no blocking, no thrown error, user can see the app, better UX

Reply:
Mmm, I'm not sure about this. This `getProvider` returns a promise: `Promise<Web3>`

This means, that if something is wrong, we can handle the case.

One solution is to do something like:
`resolve(null)` signaling, there's no Web3 object cause we didn't connect the wallet yet. The app can react and handle that case. So you can still have a good UX